jbmc/src/java_bytecode/java_types.cpp

```diff
@@ -131,6 +131,17 @@ typet java_array_element_type(const symbol_typet &array_type)
   return array_type.find_type(ID_C_element_type);
 }
 
+/// Return a const reference to the element type of a given java array type
+/// \param array_type The java array type
+const typet &java_array_element_type_ref(const symbol_typet &array_type)
+{
+  INVARIANT(
+    is_java_array_tag(array_type.get_identifier()),
+    "Symbol should have array tag");
+  return array_type.find_type(ID_C_element_type);
```

Comment:
The above is a PRECONDITION

Reply:
Shouldn't the element type be the subtype of the data member?

Reply:
There is a similar function above which returns the element type by value, I used the same format here. The benefit of invariant comparing to precondition as far as I understand it is that it provides a message, so if things go wrong it might be easier to see why. For the element type - for primitive arrays the element type is not a pointer.

Reply:
> For the element type - for primitive arrays the element type is not a pointer.

I don't really understand that comment? How do pointers enter this discussion?

Reply:
In java array, the element type is either a primitive type or a pointer to a symbol (of a non-primitive type).

Reply:
Is the "pointer to a symbol" a `symbol_typet`? Either way I don't get why that's a reason to not use `subtype()` and instead use `ID_C_element_type`? But likely I'm not understanding this properly.

Reply:
The way I understood the original comment from @kroening was to use `array_type.find_type(ID_C_element_type).subtype()`. So maybe I misunderstood then. How did you interpret it?

Reply:
Ah, mine was different: I was assuming (and that's what I do) that the need for `ID_C_element_type` is in question. That is, `array_type.subtype()` should yield whatever `array_type.find_type(ID_C_element_type)` currently yields. But maybe that's not actually a reasonable ask as there may be more information stored that I am currently thinking there is?

Reply:
Hmm, `array_type` is of type `symbol_typet:typet`, it can look something like this:
```
symbol
            * identifier: java::array[reference]
            * #base_name: array[reference]
            * #element_type: pointer
                * width: 64
                * #reference: 1
                0: symbol
                    * identifier: java::java.lang.Integer
                    * #base_name: java.lang.Integer
```
Calling `array_type.subtype()` returns `nil` in this case.

Reply:
Ah, ok, I suppose there is a reason it has to be a `symbol_typet`. Indeed a `symbol_typet` shouldn't have a `subtype()`.

Reply:
Maybe array_type is misleading because people would expect it to be an array_typet. `java_array_type` could be better?

Reply:
That's a good point, but I'm still wondering why a Java array is different from an array elsewhere.

Reply:
In general? The main distinguishing feature is they carry RTTI (i.e. you can write `Integer[] a = new Integer[5]; if(a instanceof Object[])` (which passes, as Java arrays are covariant in their element type). They can also be synchronized upon. Hence you need a control block in addition to the actual data array. I imagine the actual JVM implements that like a variable-length struct (`clsidt clsid; bool locked; element_type elements[x]`); we use two allocations, the control block (`java::array[*]`) and the variable-length data array.

Reply:
Maybe I just haven't found it: where is the API to what I'd expect to be named `java_array_typet`? Or is there a reason all of the above can't be combined into an API?

Reply:
I haven't found any either, there is only a function `java_array_type` that creates a pointer to array symbol around an element type.